### PR TITLE
Ethan/lsm keeper helpers

### PR DESCRIPTION
**Pull request by ethan-stride**
Closes: second ticket to implement helper keeper functions

## Context and purpose of the change
This update creates helper function for LSMTokenDeposits to Add instances of the same denom, to directly set a new status, and to get all deposits for a given host zone or a host zone with a particular status.  All functions unit tested

## Brief Changelog

  - *addLSMTokenDeposit, UpdateLSMTokenDepositStatus, GetLSMTokenDepositsForHostZone, GetLSMDepositsGotHostZoneWithStatus*
  - *Changed the params of the last two from the scaffold, didn't need the deposit as an input*
  - *Created unit tests on these new functions*
  

## Author's Checklist

I have...
- Run and PASSED locally all GAIA integration tests
- Added new unit tests for each new keeper function 
- Verified that the unit tests pass locally

## Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] manually tested (if applicable)
- [ ] confirmed the author wrote unit tests for new logic
- [ ] reviewed documentation exists and is accurate

## Documentation and Release Note

  - [ ] Does this pull request introduce a new feature or user-facing behavior changes? 
  - [ ] Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`?
  - [ ] This pull request updates existing proto field values (and require a backend and frontend migration)? 
  - [ ] Does this pull request change existing proto field names (and require a frontend migration)?
  How is the feature or change documented? 
      - [ ] not applicable
      - [ ] jira ticket `XXX` 
      - [ ] specification (`x/<module>/spec/`) 
      - [ ] README.md 
      - [ ] not documented

### CODE REVIEW

File: x/stakeibc/keeper/lsm_token_deposit_test.go [hunk 1]
```diff
@@ -44,9 +54,134 @@ func (s *KeeperTestSuite) TestLSMTokenDepositRemove() {
...
+		types.TRANSFER_FAILED,
+		types.DETOKENIZATION_QUEUE,
+		types.DETOKENIZATION_IN_PROGRESS,
+		types.DETOKENIZATION_FAILED,
+	}
+	deposits := s.createNLSMTokenDeposit(225) // total number across all combos
+	idx := 0
+	for hzid := 0; hzid < 5; hzid++ {
+		for sid := 0; sid < len(statuses); sid++ {
+			numCombo := (hzid + 1) * (sid + 1)
```
Comment (sampocs) [inline x/stakeibc/keeper/lsm_token_deposit_test.go:161]: I found this test a bit hard to follow, some suggestions that might clarify things a bit:
```suggestion
	// Check that we get every deposit which matches hostzone and status
	// Check that we only get deposits which match hostzone and status
	// Need a predictable, different, non-zero number of deposits for each (zone,status) combo
	numHostZones := 5
        statuses := []types.LSMDepositStatus{
		types.TRANSFER_IN_PROGRESS,
		types.TRANSFER_FAILED,
		types.DETOKENIZATION_QUEUE,
		types.DETOKENIZATION_IN_PROGRESS,
		types.DETOKENIZATION_FAILED,
	}
	
	// For each host zone + status combo, create a different number of deposits
	// that's determined by the product of (hostZone index + 1) * (status index + 1)
	// For instance:
	//    chain-0, status-0 => 1 deposit
	//    chain-0, status-1 => 1 * 2 = 2 deposit
	//    chain-1, status-1 => 2 * 2 = 4 deposits
	deposits := s.createNLSMTokenDeposit(225) // total number across all combos
	idx := 0
	for hzid := 0; hzid < numHostZones; hzid++ {
		for sid := 0; sid < len(statuses); sid++ {
			numCombo := (hzid + 1) * (sid + 1)
```

I was also trying to think if there was a way to dynamically calculate the 225 value based on the status' and host zones. For instance, the DEPOSIT_PENDING status was just added - it would be nice if this test did not have to be modified.
I couldn't think of how to do it using just an equation. Can you?

**Review by sampocs — CHANGES_REQUESTED** [review]
Fantastic work here - these tests are excellent! 

Only notable comment is, I would see if you can explain the last test a little more clearly - I found it quite confusing when reading through it for the first time.

Also, this fix is going to be super annoying, but for the `require.Equal` statements - the first parameter is the _expected_ value and the second one is the _actual_. I think you have most of these backwards.

I personally wish it was the other way around since something like:
```go
s.Require().Equal(deposit.ChainId, chainId, "Got back deposit from different chain!")
s.Require().Equal(deposit.Status, status, "Got back deposit with wrong status!")
```
looks better than:
```go
s.Require().Equal(chainId, deposit.ChainId, "Got back deposit from different chain!")
s.Require().Equal(status, deposit.Status, "Got back deposit with wrong status!")
```
But when the test fails, it's printed like:
```go
s.Require().Equal("valueA", "valueB", "some description")

// expected: valueA
// actual: valueB
// some description
```

So it can be a bit confusing to debug when these are backwards. I'd recommend just skimming through this file and making sure the order is right for each of em

File: x/stakeibc/keeper/lsm_token_deposit.go [hunk 2]
```diff
@@ -48,27 +48,52 @@ func (k Keeper) GetAllLSMTokenDeposit(ctx sdk.Context) []types.LSMTokenDeposit {
...
+	original, found := k.GetLSMTokenDeposit(ctx, deposit.ChainId, deposit.Denom)
+	if found {
+		deposit.Amount = original.Amount.Add(deposit.Amount)
+	}
+	k.SetLSMTokenDeposit(ctx, deposit)
 }
 
-func (k Keeper) UpdateLSMTokenDepositStatus(ctx sdk.Context, deposit types.LSMTokenDeposit, status types.LSMDepositStatus) {
-	// TODO [LSM]
+func (k Keeper) UpdateLSMTokenDepositStatus(
```
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/lsm_token_deposit.go:58]: It bugs me when lines are over 80 characters but that ship has long sailed in this codebase.  I think the more descriptive names are worth it, if 140 is your line length limit, now I know.  But yes I had intentionally broken lines here because it makes it easier to read the arguments at the expense of the function opening shape is not as recognizable.

Btw I also debated getting rid of this whole function because calling it is almost the same length as setting a status on the deposit and calling to SetLSMTokenDeposit instead, but whatever, it is already done and tested.

File: x/stakeibc/keeper/lsm_token_deposit_test.go [hunk 3]
```diff
@@ -4,29 +4,39 @@ import (
...
+	}
+	return deposits
+}
+
+func (s *KeeperTestSuite) createSetNLSMTokenDeposit(n int) []types.LSMTokenDeposit {
+	newDeposits := s.createNLSMTokenDeposit(n)
+	return s.setGivenLSMTokenDeposit(newDeposits)
 }
 
 func (s *KeeperTestSuite) TestLSMTokenDepositGet() {
```
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/lsm_token_deposit_test.go:38]: This was the pattern I had seen elsewhere in some of the stakeIBC code... it actually really bugged me, I would much prefer the sensible naming to keep the test as similar to the function as possible.  I'll make these changes.

File: x/stakeibc/keeper/lsm_token_deposit_test.go (hunk 1 above)
Comment (ethan-stride) [reply x/stakeibc/keeper/lsm_token_deposit_test.go:161]: yes, 15 came from the 5th triangular number T_n = 15 
and in general of course the nth triangular number T_n =  n*(n+1)/2 by the Gauss sum trick.
225 = 15^2 = T_5*T_5 and if instead we had 6 status and 5 hostzones then it would be T_6*T_5 = 21*15 = 315
Draw it as a table and in each cell is the number which is the product, but each row or column is just a multiple of the other direction's triangular number.  Visualizing it like this makes it easy to see why it is the multiplication.
I don't think this really helps make things clearer than a comment just saying this number is the total deposits needed.

File: x/stakeibc/keeper/lsm_token_deposit_test.go [hunk 4]
```diff
@@ -44,9 +54,134 @@ func (s *KeeperTestSuite) TestLSMTokenDepositRemove() {
...
+	idx := 0
+	for hzid := 0; hzid < 5; hzid++ {
+		for sid := 0; sid < len(statuses); sid++ {
+			numCombo := (hzid + 1) * (sid + 1)
+			for i := 0; i < numCombo; i++ {
+				deposits[idx].ChainId = strconv.Itoa(hzid)
+				deposits[idx].Status = statuses[sid]
+				idx++
+			}
+			s.Ctx.Logger().Info("combonum %d idx %d", numCombo, idx)
```
[the remark this answers is not in the document]
Comment (ethan-stride) [reply x/stakeibc/keeper/lsm_token_deposit_test.go:167]: good catch

File: x/stakeibc/keeper/lsm_token_deposit.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit.go]: haha gotchya, I mostly just relax the line length on the function signature. For the rest of the code, I don't target a specific number but I probably max it around 100-110. I do find that trying to force _every_ line to <80 often makes it less readable - but obviously there's a lot of personal bias there. 

Yeah it's definitely not necessary, but since we update the status often, it's just slightly more readable to be able to do it in one line. But if you think we should remove it, I'm fine with that

File: x/stakeibc/keeper/lsm_token_deposit_test.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit_test.go]: stakeibc tech debt strikes again!

[the remark this answers is not in the document]
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit_test.go]: i bet that was from ignite

File: x/stakeibc/keeper/lsm_token_deposit_test.go (hunk 1 above)
Comment (sampocs) [reply x/stakeibc/keeper/lsm_token_deposit_test.go]: ah nice - thanks for the explainer - been awhile since I've done that type of math haha

And agreed, the formula is complex enough that I don't think you need to code it up. But maybe make a note to yourself to update this test once the commit with the DEPOSIT_PENDING status is merged in

**Review by sampocs — APPROVED** [review]
Changes look great, go ahead and 🚢 !
